tests/InfillTest.cpp

```diff
@@ -0,0 +1,265 @@
+        }
+
+        friend std::ostream& operator<<(std::ostream& os, const InfillTestParameters& params)
+        {
+            return os << params.name << "(" << (params.valid ? std::string("input OK") : params.fail_reason) << ")";
+        }
+    };
+
+    constexpr coord_t outline_offset = 0;
+    constexpr coord_t infill_line_width = 350;
+    constexpr coord_t infill_overlap = 0;
+    constexpr size_t infill_multiplier = 1;
+    const AngleDegrees fill_angle = 0.;
+    constexpr coord_t z = 100; // Future improvement: Also take an uneven layer, so we get the alternate.
+    constexpr coord_t shift = 0;
+
+    InfillTestParameters generateInfillToTest(const InfillParameters& params, const size_t& test_polygon_id, const Polygons& outline_polygons)
+    {
+        const EFillMethod pattern = params.pattern;
+        const bool zig_zagify = params.zig_zagify;
+        const bool connect_polygons = params.connect_polygons;
+        const coord_t line_distance = params.line_distance;
+
+        Infill infill
+        (
+            pattern,
+            zig_zagify,
+            connect_polygons,
+            outline_polygons,
+            outline_offset,
+            infill_line_width,
+            line_distance,
+            infill_overlap,
+            infill_multiplier,
+            fill_angle,
+            z,
+            shift
+        ); // There are some optional parameters, but these will do for now (future improvement?).
+
+        Polygons result_polygons;
+        Polygons result_lines;
+        infill.generate(result_polygons, result_lines, nullptr, nullptr);
+
+        InfillTestParameters result = InfillTestParameters(params, test_polygon_id, outline_polygons, result_lines, result_polygons);
+        return result;
+    }
+
+    std::vector<InfillTestParameters> generateInfillTests()
+    {
+        std::vector<Polygons> shapes;
+        if (!readTestPolygons("../tests/polygons.txt", shapes))
+        {
+            return { InfillTestParameters() };  // return an invalid class, that'll trip up the 'file read' assertion in the TEST_P's
+        }
+
+        std::vector<EFillMethod> methods;
+        std::vector<EFillMethod> skip_methods = { EFillMethod::CROSS, EFillMethod::CROSS_3D, EFillMethod::CUBICSUBDIV }; //TODO: Support for testing infill that needs the sierpinski-provider!
+        for (int i_method = 0; i_method < static_cast<int>(EFillMethod::NONE); ++i_method)
+        {
+            const EFillMethod method = static_cast<EFillMethod>(i_method);
+            if (std::find(skip_methods.begin(), skip_methods.end(), method) == skip_methods.end()) // Only use if not in skipped.
+            {
+                methods.push_back(method);
+            }
+        }
+
+        std::vector<coord_t> line_distances = { 400, 600, 800, 1200 }; // TODO?: Gyroid fails the 'fill less than 100% of area available' with values close to the line width, like 350.
```

Comment:
When searching for skipping parameterized tests, I only found [this](https://stackoverflow.com/questions/25999057/how-to-signal-to-gtest-that-a-test-wants-to-skip-itself) unacceptable (but funny) solution.

Also keep in mind that the 350 line distance for lines that are 350 wide would amount to 100% infill, which is a case we avoid anyway (in the front-end if I'm not mistaken).

So I think we can merge this comment with the one you made about `... << "Infill area should be within 15% of expected size."; // TODO: 10 ~ 20% is actually quite bad?`.

Reply:
There is a neater way. You can call `GTEST_SKIP()` inside the test.

However when I tried to apply that to your test, I found a bit of a problem that I overlooked before. You have a function that generates the parameter list of the test, but in that function you actually run the function under test.

If we move the function under test into the actual `TEST_P`, then we can check there for `line_distance` to be <= 400 and the pattern to be Gyroid. This move would be better anyway because `INSTANTIATE_TEST_CASE_P` should provide the parameters for the test, not the result of the test. Because it provides the parameters of the test, we can actually check for certain parameter combinations and skip those. We can't skip the generating of the parameters.

Reply:
I think I still got an old version of gtest, GTEST_SKIP may have been added to a version after my install, I'll have to do this next week.

Reply:
We've already discussed this, but for people following along & for future reference:

It turns out `GTEST_SKIP` is only available in their master, not in the latest stable (1.8.1). Rather than rely on an unstable version, we've just decided to keep it mostly 'as-is' for now. We should update if they've got a new version out, but that could take a while.